zend_abstract_interface/functions/php8/functions.c

```diff
@@ -0,0 +1,116 @@
+#include <zai_assert/zai_assert.h>
+
+bool zai_call_function(const char *name, size_t name_len, zval *retval, int argc, ...) {
+    if (!retval) return false;
+
+    /* For consistency with zend_call_function(), this wrapper also initializes
+     * the retval to IS_UNDEF so that the caller can expect an undefined retval
+     * in the error cases. Past wrappers of zend_call_function() did not mimic
+     * this behavior and that lead to a SIGSEGV.
+     *
+     * https://github.com/php/php-src/blob/PHP-8.0.3/Zend/zend_execute_API.c#L673
+     */
+    ZVAL_UNDEF(retval);
+
+    if (!name || !name_len || argc < 0) return false;
+
+    /* Functions cannot be called outside of a request context.
+     * PG(modules_activated) indicates that all of the module RINITs have been
+     * called and we are in a request context.
+     */
+    if (!PG(modules_activated)) return false;
+
+    zai_assert_is_lower(name, "Function names must be lowercase.");
+    assert(*name != '\\' && "Function names must not have a root scope prefix '\\'.");
+
+    zend_function *func = NULL;
+    zend_string *zsname = NULL;
+    ALLOCA_FLAG(use_heap)
+
+    /* We look up the function handler directly from zend_fetch_function()
+     * instead of calling zend_is_callable_ex() because:
+     *
+     *   - We always call functions by their string name and not one of the
+     *     many 'callable' types, so the code path can be simplified.
+     *   - We assume the caller always provides the lowercased function name so
+     *     we can avoid zend_str_tolower_copy().
+     *   - We assume the caller will not prefix the function name with a
+     *     root-scope prefix, '\'.
+     *
+     * In addition, we look up the function handler from zend_fetch_function()
+     * instead of directly from the EG(function_table) to ensure that the
+     * runtime cache for userland functions will be initialized.
+     */
+    ZSTR_ALLOCA_ALLOC(zsname, name_len, use_heap);
+    memcpy(ZSTR_VAL(zsname), name, (name_len + 1));
+    func = zend_fetch_function(zsname);
+    ZSTR_ALLOCA_FREE(zsname, use_heap);
+
+    /* "function %s() does not exist" */
+    if (!func) return false;
+
+    zend_fcall_info_cache fcc = {0};
+    fcc.function_handler = func;
+
+    zend_fcall_info fci = {0};
+    fci.size = sizeof(zend_fcall_info);
+    fci.retval = retval;
+
+    /* We add the zval args directly from the variable arguments API instead of
+     * using zend_fcall_info_argv() which allows us to NULL-check each arg. It
+     * also allows us to avoid an unnecessary call to
+     * zend_fcall_info_args_clear() since the 'param{s|_count}' members in
+     * 'zend_fcall_info' will always be zeroed out in our case.
+     */
+    zval params[argc];
+    if (argc > 0) {
+        va_list argv;
+        va_start(argv, argc);
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            zval *arg = va_arg(argv, zval *);
+            if (!arg) return false;
+            params[i] = *arg;
```

Comment:
The canonical way to assign a zval would be actually `ZVAL_COPY_VALUE()`. This copies the zend_value as two individual 32 bit assignments on 32 bit environments (I do not know why. Perhaps that's completely irrelevant?!?) and notably does _not_ copy u2.

Reply:
TIL! Addressed in 472d6593. Thanks for the insight.